src/citrine/_rest/collection.py

```diff
@@ -98,6 +104,70 @@ def list(self,
                 # properly, so we are filtering client-side.
                 pass
 
+    def list(self,
+             page: Optional[int] = None,
+             per_page: Optional[int] = None) -> Iterable[ResourceType]:
+        """
+        List all visible elements in the collection.
+
+        Leaving page and per_page as default values will yield all elements in the
+        collection, paginating over all available pages.
+
+        Parameters
+        ---------
+        page: int, optional
+            The "page" of results to list. Default is to read all pages and yield
+            all results.
+        per_page: int, optional
+            Max number of results to return per page. Default is 20.  If the page
+            parameter is specified this will limit the number of results in that
+            response.
+
+        Returns
+        -------
+        Iterable[ResourceType]
+            Resources in this collection.
+
+        """
+        # Do an initial fetch of the first page using supplied args.
+        # If more results expected, proceed with paginated calls.
+        first_page = self._fetch_page(page=page, per_page=per_page)
+        first_page_count = 0
+        first_uid = None
+        for idx, element in enumerate(first_page):
+            yield element
+
+            uid = getattr(element, 'uid', None)
+            if idx == 0 and not first_uid and uid:
+                first_uid = uid
+
+            first_page_count += 1
+
+        if page is not None or first_page_count != (per_page or DEFAULT_PER_PAGE):
+            return
+
+        # Read extra pages until nothing is left to read
+        next_page = 2
+        while True:
+            subset = self._fetch_page(page=next_page, per_page=per_page)
+            count = 0
+            for idx, element in enumerate(subset):
+                # escaping from infinite loops where page/per_page are not
+                # honored and are returning the same results regardless of page:
+                uid = getattr(element, 'uid', None)
+                if first_uid is not None and first_uid == uid:
+                    break
```

Comment:
Should we throw an exception here? Is this ever valid server-side behavior?

Reply:
Previously the server was always returning the last page, leading to infinite loops. If there were 40 elements and per_page=20, then page 1 returns elements1-20, page 2 returns elements 21-40, and page 3 _also_ returns elements 21-40. It's possible that your work on paging also fixed this behavior for v1 routes, but if not, then some check like this will be necessary.

However, I do not think this logic will solve the above problem. If I understand correctly, this checks to see if the very first uid ever gets returned again. But the problem (which may have changed) wasn't that it looped back to the very beginning of the list, but that it looped back to the beginning of the last page.

Reply:
I didn't know about the problem you mentioned above.  I have seen an endpoint that completely ignores the page/per_page params and returns all items on every call so was trying to detect that case.

I think there's a relatively easy way to modify this to support both cases.  Instead of checking against the first UID we could check the first element in the previous list of elements.  I'll implement that now.  For the time being I would like to avoid raising an exception so we can get the end to end tests passing.  Once I fix the misbehaving endpoint I think we should raise an exception if this behavior is detected.

I also came to the conclusion that this cursor based pagination won't work for the v1 page/per_page paginated endpoints.  I think this makes this PR necessary.  Once we fully support cursor based pagination in the API we can swap over.